File: csharp/DesiArchitect.DesignPatterns/04_Decorator/After/DeliveryDecorator.cs
namespace DesiArchitect.DesignPatterns.Decorator.After;

public class DeliveryDecorator : IBill
{
    private readonly IBill _inner;
    private readonly decimal _charge = 40m;

    public DeliveryDecorator(IBill inner) => _inner = inner;

    public decimal Total()
    {
        var subtotal = _inner.Total();
        var result = subtotal + _charge;
        Console.WriteLine($"  🚚 + Delivery: ₹{_charge} -> ₹{result}");
        return result;
    }
}

File: typescript/src/04-decorator/after/delivery-decorator.ts
import type { IBill } from './ibill.js';

export class DeliveryDecorator implements IBill {
  private readonly charge = 40;

  constructor(private readonly inner: IBill) {}

  total(): number {
    const subtotal = this.inner.total();
    const result = subtotal + this.charge;
    console.log(`  🚚 + Delivery: ₹${this.charge} -> ₹${result}`);
    return result;
  }
}